File: LintCode-Java/ValidateBinarySearchTree.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class ValidateBinarySearchTree {
    /**
     * @param root: The root of binary tree.
     * @return: True if the binary tree is BST, or false
     */

    // V1, O(n)
    // Divide&Conquer
    // BST:
    // 1. Left tree is BST;
    // 2. Right tree is BST;
    // 3. max value of left subtree < root value < min value of the right sub tree.
    public boolean isValidBST(TreeNode root) {
        Result result = helper(root);
        return result.isBST;
    }

    private Result helper(TreeNode root) {
        Result result = new Result(true, Integer.MAX_VALUE, Integer.MIN_VALUE);
        if (root == null) {
            return result;
        }

        // Divide
        Result left = helper(root.left);
        Result right = helper(root.right);

        // Conquer
        if (!left.isBST || !right.isBST) {
            result.isBST = false;
            return result;
        }
        if (root.left != null && root.val <= left.max) {
            result.isBST = false;
            return result;
        }
        if (root.right != null && root.val >= right.min) {
            result.isBST = false;
            return result;
        }
        return new Result(true,
                Math.min(root.val, left.min),
                Math.max(root.val, right.max));
    }

    private class Result {
        boolean isBST;
        int min, max;
        Result (boolean isBST, int min, int max) {
            this.isBST = isBST;
            this.min = min;
            this.max = max;
        }
    }

    // V2, O(n)
    // Inorder Traversal is ascending <=> BST
    private Integer lastVal = null;
    public boolean isValidBST(TreeNode root) {
        if (root == null) {
            return true;
        }

        if (!isValidBST(root.left)) {
            return false;
        }

        if (lastVal != null && root.val <= lastVal) {
            return false;
        }
        lastVal = root.val;

        if (!isValidBST(root.right)) {
            return false;
        }

        return true;
    }
}

File: LintCode/python/validate_binary_search_tree.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param root: The root of binary tree.
    @return: True if the binary tree is BST, or false
    """
    def isValidBST1(self, root):
        # [divide and conquer]
        # time:  O(n)
        # space: O(1)
        if not root:
            return True

        isBST, _, __ = self.helper(root)
        return isBST

    def helper(self, root):
        if not root:
            return (True, sys.maxint, -sys.maxint)

        # Divide
        isLeftBST, leftMin, leftMax = self.helper(root.left)
        isRightBST, rightMin, rightMax = self.helper(root.right)

        # Conquer
        return (isLeftBST and isRightBST and leftMax < root.val and root.val < rightMin,
                min(leftMin, root.val),
                max(rightMax, root.val))

    prevNode = None
    def isValidBST2(self, root):
        # [dfs]
        # time:  O(n)
        # space: O(1)
        # inorder traversal is ascending
        if not root:
            return True

        if not self.isValidBST(root.left):
            return False

        if self.prevNode and self.prevNode.val >= root.val:
            return False
        self.prevNode = root

        if not self.isValidBST(root.right):
            return False

        return True

